### DISCUSSION TITLE
Automatic nonce handling for DryocSecretBox

### DISCUSSION THREAD

**Discussion by kw217** [brndnmtthws/dryoc#37] [Ideas]
I want to use symmetric encryption for some database columns. `DryocSecretBox` is clearly the right way to go, but it's not immediately obvious how to handle the nonce. Either I need separate columns for the nonces (one for each column I want to encrypt), or I need to store the nonce and ciphertext in the same column in some appropriate format. It would be great if dryoc did that for me.

I notice that orion has a mode where it both automatically generates the nonce (so you can't accidentally mismanage it, e.g., by reusing) and prepends it (so you don't have to figure out how to store and retrieve it): [orion::aead](https://docs.rs/orion/latest/orion/aead/index.html).

Would this be a useful addition to dryoc, and what would be the best way to add it? I can imagine a `DryocEasySecretBox` which has a nonce field, but there are quite a few new names to be thought of. An alternative would be to make a breaking change and always store the nonce (dryoc could still retain a method to supply a nonce on creation, but add a new one which picks it automatically).

**Comment by brndnmtthws** [comment]
Nonce handling _is_ a bit tricky, but here are a few rules you can follow to avoid problems:

- Don't reuse the same nonce: each time you encrypt, make a new nonce (common practice is to increment a sequence number)
- Nonces can be safely stored alongside an encrypted message, or transmitted over the wire, they themselves are not secrets, and knowing the value of a nonce doesn't reveal information about the content of the message
- In the case of storing encrypted messages in a database: you _wouldn't_ want to use a row ID (such as a primary key, for example) as a nonce, instead you should put the nonce in a separate column, or store it alongside the message payload, so that you can update the nonce every time the message is updated
- If you're storing encrypted messages in a DB, every time you re-encrypt the message (the U in CRUD), you would generate a new nonce, which could be random number or sequential, but not something used recently

I could provide a basic sequential nonce (i.e., a number that just automatically increments starting from a random value), but I wouldn't want to require that you use the one I provide, because you can also take advantage of a nonce to improve security in certain cases.

For example in a network API protocol: the nonce can be selected once during the initial handshake, and then sequentially incremented by the client and server separately each time a message is exchanged, but after the initial handshake you would never transmit the nonce over the wire again. This way, if someone intercepts some of the messages, it would be nearly impossible to figure out the value of the nonce unless the spy has a way to guarantee they've intercepted 100% of all the messages exchanged since the beginning of time, including the nonce. It wouldn't be possible to know the value of the nonce if you miss even just one message. This is a pretty common pattern for handling nonces within well-designed protocols.

Sequences are commonly used because of the aforementioned use case: once the two parties involved know the initial nonce, you can simply increment it forever and you don't need to transmit the nonce over the wire. Alice and Bob only need to know the nonce at the beginning of a protocol. Of course, if one of the parties misses a message or loses track of the nonce, you would have to do a new handshake.

Sequences are good because they make it hard to accidentally reuse the same value, which helps protect you from replay attacks. In the case of a 64 bit nonce, you have $2^{64}-1$ possible values before you reuse one, which is a very big number.

For the reasons mentioned above, I would consider it bad practice to enforce a certain behaviour or pattern for handling nonces, but I could still include a basic sequential nonce in the library for convenience, and leave it up to individuals to decide whether to use it or not, without requiring it.

**Reply by kw217** [reply]
Thanks; my concern is less with how to set the nonce (though as you say there are good practice ways to do that and it would be nice if the library encouraged clients in that direction), and more with how to store/transmit the nonce. A common pattern is to prepend it to the encrypted message; it would be nice if that was built in function rather than requiring the client to always do a vector append.

**Reply by brndnmtthws** [reply]
Here's a simple example with serde showing how you can serialize to JSON or MsgPack: https://github.com/brndnmtthws/box-with-nonce

It's as simple as doing the following:

```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct BoxWithNonce<Nonce, Box> {
    nonce: Nonce,
    message: Box,
}

impl<Nonce, Box> BoxWithNonce<Nonce, Box> {
    pub fn new(nonce: Nonce, message: Box) -> Self {
        Self { nonce, message }
    }
}
```

And then:

```rust
// Encrypt `message`, into a Vec-based box
let box_with_nonce = BoxWithNonce::new(
    nonce.clone(),
    DryocSecretBox::encrypt_to_vecbox(message, &nonce, &secret_key),
);
```

Now your nonce and message can live happily ever after side by side.

**Reply by kw217** [reply]
Sure, that works. In my code I'm just writing the (fixed-length) nonce into the vector first, and then appending the ciphertext; on read I use `split_at` to take off the first N bytes and use that as the nonce, then decrypt the remainder. There are clearly various ways of doing this - it would be nice if dryoc was opinionated and just handled it for me, rather than leaving me to come up with a scheme.

**Reply by SimenB** [reply]
I hit something similar now. We also prepend the nonce, so I have the following code:

```rust
if token.len() < CRYPTO_SECRETBOX_NONCEBYTES {
    return Err(miette!(
        "token must contain a nonce and a message (must be larger than {} bytes, is {})",
        CRYPTO_SECRETBOX_NONCEBYTES,
        token.len(),
    ));
}

let (nonce, message) = token.split_at(CRYPTO_SECRETBOX_NONCEBYTES);

let nonce = Nonce::try_from(nonce).unwrap();

let secret_box = VecBox::from_bytes(message).unwrap();

let message = secret_box.decrypt_to_vec(&nonce, &secret_key).unwrap();
```

It would be nice if there was some sort of convenience API so I could do something like the following:

```rust
let secret_box = VecNonceBox::from_bytes_with_nonce(token).unwrap();

let message = secret_box.decrypt_to_vec(&secret_key).unwrap();
```

or some such 🙂 

---

Anyways, thank you so much for creating this library - being able to first port my existing code using libsodium (from Go and TS) more or less directly, then use the higher level abstractions afterwards for more idiomatic code was a very pleasant experience! 👍
